app/models/forms/export.rb

```diff
@@ -21,6 +60,117 @@ def to_csv
       end
     end
 
+    # rubocop:disable Metrics/MethodLength, Metrics/BlockLength, Metrics/AbcSize, Metrics/PerceivedComplexity, Style/Next
+    def to_xls
+      # TODO: option set "levels"?
+
+      book = Spreadsheet::Workbook.new
+
+      # Create sheets
+      questions = book.create_worksheet(name: "survey")
+      choices = book.create_worksheet(name: "choices")
+      settings = book.create_worksheet(name: "settings")
+
+      # Write sheet headings at row index 0
+      questions.row(0).push("type", "name", "label", "required", "relevant", "constraint")
+      choices.row(0).push("list_name", "name", "label")
+      settings.row(0).push("form_title", "form_id", "version", "default_language")
+
+      group_depth = 1 # assume base level
+      repeat_depth = 1
+      index_mod = 1 # start at row index 1
+      choices_index_mod = 0
+      option_sets_used = []
+
+      @form.preordered_items.each_with_index do |q, i|
+        # did one or more groups just end?
+        # if so, the qing's depth will be smaller than the depth counter
+        while group_depth > q.ancestry_depth
+          # are we in a repeat group?
+          # we don't want to end the repeat if we are ending a nested non-repeat group within a repeat
+          if repeat_depth > 1 && repeat_depth >= group_depth
+            questions.row(i + index_mod).push("end repeat")
+            repeat_depth -= 1
+          else
+            # end the group
+            questions.row(i + index_mod).push("end group")
+          end
+
+          # update counters
+          group_depth -= 1
+          index_mod += 1
+        end
+
+        if q.group? # is this a group?
+          if q.repeatable?
+            questions.row(i + index_mod).push("begin repeat", q.code)
+            repeat_depth += 1
+          else
+            questions.row(i + index_mod).push("begin group", q.code)
+          end
+
+          # update counters
+          group_depth += 1
+        else # is this a question?
+          # do we have an option set?
+          if q.option_set_id.present?
+            os = OptionSet.find(q.option_set_id)
+            os_name = " #{os.name}"
+            os_already_logged = option_sets_used.include?(q.option_set_id)
+
+            # log the option set to the spreadsheet if we haven't yet
+            unless os_already_logged
+              os.option_nodes.each_with_index do |node, x|
+                if node.option.present?
+                  choices
+                    .row(x + choices_index_mod)
```

Comment:
what is `x`, an index? please name it more verbosely to help with readability since things are deeply nested here